Make it work

dzhlobo opened the pull request:
We made this pull request on a mob-programming session with awesome Rubizza 2019 students.

![photo_2019-08-23 19 00 08](https://user-images.githubusercontent.com/800337/63606241-47387180-c5d8-11e9-8a19-b22f4fcd09fc.jpeg)
![photo_2019-08-23 19 00 16](https://user-images.githubusercontent.com/800337/63606245-49023500-c5d8-11e9-869c-6156f5ac3ffe.jpeg)

dzhlobo committed:
Show the correct licence in README

dzhlobo committed:
Setup test environment

dzhlobo committed:
Add code examples

dzhlobo committed:
Make CryptedHash work as Hash

Also we can serialize it and save to a file and restore it from the
file.

Co-authored-by: Artem Starovoitov <epfeen@gmail.com>
Co-authored-by: Maxim Spaseonav <maxrikal@gmail.com>
Co-authored-by: Ilya Vasilevich <inudad@yandex.ru>
Co-authored-by: Alexey Varvashevich <Amix047@gmail.com>
Co-authored-by: Ivan Barysiuk <Jdjjad89coq5ven5uqoweOOvqw3n@gmail.com>
Co-authored-by: Alex Samets <aliaksandr.samets@gmail.com>
Co-authored-by: Margarita Omelchuk <margosha20617@gmail.com>
Co-authored-by: Sokurenko Mikhail <tedpoolioo@gmail.com>
Co-authored-by: Rudy Anton <midda7@mail.ru>
Co-authored-by: Alexander Baranouski <Fr08ty@gmail.com>
Co-authored-by: Julia Lugina <lugina.yulya@mail.ru>
Co-authored-by: Andrew Berendeev <soligorsk162@gmail.com>
Co-authored-by: Kirill Loktevich <lokkirill@gmail.com>

dzhlobo commented:
@Maxxx911, @IvanBoris, you mentioned you want to continue the development. I think we should add encryption logic before merging this. I've added you as collaborators to my fork https://github.com/dzhlobo/crypted_hash so you can push your code directly to the `make-it-useful` branch to my repository and your commits will appear in this pull request.

Here's an example of simple encryption in Ruby: https://stackoverflow.com/a/4129215/703860.

avsej commented on crypted_hash.gemspec:
```diff
@@ -9,5 +9,14 @@ Gem::Specification.new do |s|
   s.homepage    =
     'http://rubygems.org/gems/crypted_hash'
-  s.license       = 'Apache License'
+  s.license     = 'Apache License'
```
Это неправильный идентификатор лицензии. Должно быть `Apache-2.0`

https://spdx.org/licenses/
https://guides.rubygems.org/specification-reference/#license=

dzhlobo replied:
@avsej, thank you, I'll fix it.

avsej commented on lib/crypted_hash.rb:
```diff
@@ -1,2 +1,27 @@
 class CryptedHash
+  attr_reader :key, :hash
+  def initialize(key:, default_hash: {})
+    @key = key
```
лучше не использовать `key` тут, чтобы не путать с ключом ниже. Как насчёт `secret`?

dzhlobo replied:
Agree 👍.

avsej commented on spec/crypted_hash_spec.rb:
```diff
@@ -0,0 +1,43 @@
+    FileUtils.rm('./tmp/test')
+    hash.save!('./tmp/test')
+
+    expect(File.exists?('./tmp/test')).to eq(true)
```
а так не сработает?
```
expect(File).to be_exist('./tmp/test')
```

dzhlobo replied:
It will work but "be exist" sounds weird so I'm leaving it as is.

dzhlobo committed:
Fix license id

See https://guides.rubygems.org/specification-reference/#license=

dzhlobo committed:
Reduce verboseness of CryptedHash interface

Make it easier and less verbose to use it. We no longer use keywork arguments for initializing crypted hash

dzhlobo committed:
Test that we can prepopulate CryptedHash

We introduced prepopulation to initialize hash when we restore it from a
file so we have it and it is a good idea to test it as well.

dzhlobo committed:
Add Cipher class responsible for encryption

dzhlobo committed:
Encrypt hash data before saving it to a file

dzhlobo committed:
Refactor tests

Create tempfile with random name so we can run tests in parallel.

dzhlobo committed:
Raise meaningful error when we can't load a file

dzhlobo commented:
I've added encryption logic and I think it's ready to be merged.

avsej commented on lib/crypted_hash.rb:
```diff
@@ -1,2 +1,34 @@
 class CryptedHash
+  class DecryptionError < Exception; end
+
+  attr_reader :secret, :hash
```
Why `@secret` and `@hash` are publicly visible? It should not be in the API at all.
